In [5]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np
import math

## Importing the Data

In [6]:
data = pd.read_csv(r'C:\Python\PythonProject-1\Python Project - Bank Lending\XYZCorp_LendingData.txt',encoding = 'utf-8', sep = '\t')

C:\Users\ALI\Anaconda3\lib\site-packages\IPython\core\interactiveshell.py:3049: DtypeWarning: Columns (17,45,53) have mixed types. Specify dtype option on import or set low_memory=False.
  interactivity=interactivity, compiler=compiler, result=result)


In [7]:
data.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599     5000.0       5000.0           4975.0   36 months   
1  1077430    1314167     2500.0       2500.0           2500.0   60 months   
2  1077175    1313524     2400.0       2400.0           2400.0   36 months   
3  1076863    1277178    10000.0      10000.0          10000.0   36 months   
4  1075358    1311748     3000.0       3000.0           3000.0   60 months   

   int_rate  installment grade sub_grade  ... il_util open_rv_12m open_rv_24m  \
0     10.65       162.87     B        B2  ...     NaN         NaN         NaN   
1     15.27        59.83     C        C4  ...     NaN         NaN         NaN   
2     15.96        84.33     C        C5  ...     NaN         NaN         NaN   
3     13.49       339.31     C        C1  ...     NaN         NaN         NaN   
4     12.69        67.79     B        B5  ...     NaN         NaN         NaN   

   max_bal_bc all_util total_rev_hi_lim inq_fi total_cu_tl inq_last_12m  \
0         NaN      NaN              NaN    NaN         NaN          NaN   
1         NaN      NaN              NaN    NaN         NaN          NaN   
2         NaN      NaN              NaN    NaN         NaN          NaN   
3         NaN      NaN              NaN    NaN         NaN          NaN   
4         NaN      NaN              NaN    NaN         NaN          NaN   

  default_ind  
0           0  
1           1  
2           0  
3           0  
4           0  

[5 rows x 73 columns]

In [8]:
data.dtypes

id                               int64
member_id                        int64
loan_amnt                      float64
funded_amnt                    float64
funded_amnt_inv                float64
term                            object
int_rate                       float64
installment                    float64
grade                           object
sub_grade                       object
emp_title                       object
emp_length                      object
home_ownership                  object
annual_inc                     float64
verification_status             object
issue_d                         object
pymnt_plan                      object
desc                            object
purpose                         object
title                           object
zip_code                        object
addr_state                      object
dti                            float64
delinq_2yrs                    float64
earliest_cr_line                object
inq_last_6mths           

In [9]:
#create a copy of the data frame to revert back to changes

df=pd.DataFrame.copy(data)

In [10]:
df.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599     5000.0       5000.0           4975.0   36 months   
1  1077430    1314167     2500.0       2500.0           2500.0   60 months   
2  1077175    1313524     2400.0       2400.0           2400.0   36 months   
3  1076863    1277178    10000.0      10000.0          10000.0   36 months   
4  1075358    1311748     3000.0       3000.0           3000.0   60 months   

   int_rate  installment grade sub_grade  ... il_util open_rv_12m open_rv_24m  \
0     10.65       162.87     B        B2  ...     NaN         NaN         NaN   
1     15.27        59.83     C        C4  ...     NaN         NaN         NaN   
2     15.96        84.33     C        C5  ...     NaN         NaN         NaN   
3     13.49       339.31     C        C1  ...     NaN         NaN         NaN   
4     12.69        67.79     B        B5  ...     NaN         NaN         NaN   

   max_bal_bc all_util total_rev_hi_lim inq_fi total_cu_tl inq_last_12m  \
0         NaN      NaN              NaN    NaN         NaN          NaN   
1         NaN      NaN              NaN    NaN         NaN          NaN   
2         NaN      NaN              NaN    NaN         NaN          NaN   
3         NaN      NaN              NaN    NaN         NaN          NaN   
4         NaN      NaN              NaN    NaN         NaN          NaN   

  default_ind  
0           0  
1           1  
2           0  
3           0  
4           0  

[5 rows x 73 columns]

In [11]:
df.shape

(855969, 73)

In [12]:
df.dtypes

id                               int64
member_id                        int64
loan_amnt                      float64
funded_amnt                    float64
funded_amnt_inv                float64
term                            object
int_rate                       float64
installment                    float64
grade                           object
sub_grade                       object
emp_title                       object
emp_length                      object
home_ownership                  object
annual_inc                     float64
verification_status             object
issue_d                         object
pymnt_plan                      object
desc                            object
purpose                         object
title                           object
zip_code                        object
addr_state                      object
dti                            float64
delinq_2yrs                    float64
earliest_cr_line                object
inq_last_6mths           

In [13]:
data.dtypes

id                               int64
member_id                        int64
loan_amnt                      float64
funded_amnt                    float64
funded_amnt_inv                float64
term                            object
int_rate                       float64
installment                    float64
grade                           object
sub_grade                       object
emp_title                       object
emp_length                      object
home_ownership                  object
annual_inc                     float64
verification_status             object
issue_d                         object
pymnt_plan                      object
desc                            object
purpose                         object
title                           object
zip_code                        object
addr_state                      object
dti                            float64
delinq_2yrs                    float64
earliest_cr_line                object
inq_last_6mths           

## Data Cleaning

In [14]:
#check for null values

df.isnull().sum()

id                                  0
member_id                           0
loan_amnt                           0
funded_amnt                         0
funded_amnt_inv                     0
term                                0
int_rate                            0
installment                         0
grade                               0
sub_grade                           0
emp_title                       49443
emp_length                      43061
home_ownership                      0
annual_inc                          0
verification_status                 0
issue_d                             0
pymnt_plan                          0
desc                           734157
purpose                             0
title                              33
zip_code                            0
addr_state                          0
dti                                 0
delinq_2yrs                         0
earliest_cr_line                    0
inq_last_6mths                      0
mths_since_l

In [15]:
#finding variables with more than 75% missing values

null_val = df.isnull().sum(axis=0).sort_values( ascending=False)/float(len(data) )
null_val[ null_val > 0.75 ]

dti_joint                      0.999486
verification_status_joint      0.999484
annual_inc_joint               0.999484
il_util                        0.986438
mths_since_rcnt_il             0.984890
total_bal_il                   0.984476
inq_last_12m                   0.984476
open_acc_6m                    0.984476
open_il_6m                     0.984476
open_il_24m                    0.984476
open_il_12m                    0.984476
open_rv_12m                    0.984476
open_rv_24m                    0.984476
max_bal_bc                     0.984476
all_util                       0.984476
inq_fi                         0.984476
total_cu_tl                    0.984476
desc                           0.857691
mths_since_last_record         0.846742
mths_since_last_major_derog    0.750997
dtype: float64

Some columns have missing values more than 75% when compared to the original values.
So we will take 0.75 as threshold and eliminate all the columns with more than 75% missing values

In [16]:
df.drop( null_val[ null_val > 0.75 ].index, axis = 1, inplace = True ) 

In [17]:
df.isnull().sum()

id                                 0
member_id                          0
loan_amnt                          0
funded_amnt                        0
funded_amnt_inv                    0
term                               0
int_rate                           0
installment                        0
grade                              0
sub_grade                          0
emp_title                      49443
emp_length                     43061
home_ownership                     0
annual_inc                         0
verification_status                0
issue_d                            0
pymnt_plan                         0
purpose                            0
title                             33
zip_code                           0
addr_state                         0
dti                                0
delinq_2yrs                        0
earliest_cr_line                   0
inq_last_6mths                     0
mths_since_last_delinq        439812
open_acc                           0
p

In [18]:
df.shape

(855969, 53)

**As we can see that original number of observations : (855969, 73)**

**Observations after removing columns with more than 75% missing values : (855969, 53)**

**We can see that 20 columns were removed**


In [19]:
df.isnull().sum()

id                                 0
member_id                          0
loan_amnt                          0
funded_amnt                        0
funded_amnt_inv                    0
term                               0
int_rate                           0
installment                        0
grade                              0
sub_grade                          0
emp_title                      49443
emp_length                     43061
home_ownership                     0
annual_inc                         0
verification_status                0
issue_d                            0
pymnt_plan                         0
purpose                            0
title                             33
zip_code                           0
addr_state                         0
dti                                0
delinq_2yrs                        0
earliest_cr_line                   0
inq_last_6mths                     0
mths_since_last_delinq        439812
open_acc                           0
p

## Generic Code to impute the missing values
**We use Mode in case of 'Object Variables' and Mean in case of 'Numeric Variable' to impute missing values**

In [20]:
for x in df.columns[:]:
    if df[x].dtype=='object':
        df[x].fillna(df[x].mode()[0],inplace=True)
    elif df[x].dtype=='int64' or df[x].dtype=='float64':
        df[x].fillna(df[x].mean(),inplace=True)

In [21]:
df.isnull().sum()

id                            0
member_id                     0
loan_amnt                     0
funded_amnt                   0
funded_amnt_inv               0
term                          0
int_rate                      0
installment                   0
grade                         0
sub_grade                     0
emp_title                     0
emp_length                    0
home_ownership                0
annual_inc                    0
verification_status           0
issue_d                       0
pymnt_plan                    0
purpose                       0
title                         0
zip_code                      0
addr_state                    0
dti                           0
delinq_2yrs                   0
earliest_cr_line              0
inq_last_6mths                0
mths_since_last_delinq        0
open_acc                      0
pub_rec                       0
revol_bal                     0
revol_util                    0
total_acc                     0
initial_

In [22]:
df.columns


Index(['id', 'member_id', 'loan_amnt', 'funded_amnt', 'funded_amnt_inv',
       'term', 'int_rate', 'installment', 'grade', 'sub_grade', 'emp_title',
       'emp_length', 'home_ownership', 'annual_inc', 'verification_status',
       'issue_d', 'pymnt_plan', 'purpose', 'title', 'zip_code', 'addr_state',
       'dti', 'delinq_2yrs', 'earliest_cr_line', 'inq_last_6mths',
       'mths_since_last_delinq', 'open_acc', 'pub_rec', 'revol_bal',
       'revol_util', 'total_acc', 'initial_list_status', 'out_prncp',
       'out_prncp_inv', 'total_pymnt', 'total_pymnt_inv', 'total_rec_prncp',
       'total_rec_int', 'total_rec_late_fee', 'recoveries',
       'collection_recovery_fee', 'last_pymnt_d', 'last_pymnt_amnt',
       'next_pymnt_d', 'last_credit_pull_d', 'collections_12_mths_ex_med',
       'policy_code', 'application_type', 'acc_now_delinq', 'tot_coll_amt',
       'tot_cur_bal', 'total_rev_hi_lim', 'default_ind'],
      dtype='object')

In [23]:

df.head()

id  member_id  loan_amnt  funded_amnt  funded_amnt_inv        term  \
0  1077501    1296599     5000.0       5000.0           4975.0   36 months   
1  1077430    1314167     2500.0       2500.0           2500.0   60 months   
2  1077175    1313524     2400.0       2400.0           2400.0   36 months   
3  1076863    1277178    10000.0      10000.0          10000.0   36 months   
4  1075358    1311748     3000.0       3000.0           3000.0   60 months   

   int_rate  installment grade sub_grade  ... next_pymnt_d last_credit_pull_d  \
0     10.65       162.87     B        B2  ...     Feb-2016           Jan-2016   
1     15.27        59.83     C        C4  ...     Feb-2016           Sep-2013   
2     15.96        84.33     C        C5  ...     Feb-2016           Jan-2016   
3     13.49       339.31     C        C1  ...     Feb-2016           Jan-2015   
4     12.69        67.79     B        B5  ...     Feb-2016           Jan-2016   

  collections_12_mths_ex_med  policy_code application_type acc_now_delinq  \
0                        0.0          1.0       INDIVIDUAL            0.0   
1                        0.0          1.0       INDIVIDUAL            0.0   
2                        0.0          1.0       INDIVIDUAL            0.0   
3                        0.0          1.0       INDIVIDUAL            0.0   
4                        0.0          1.0       INDIVIDUAL            0.0   

  tot_coll_amt    tot_cur_bal total_rev_hi_lim default_ind  
0   225.412882  139766.247529     32163.574526           0  
1   225.412882  139766.247529     32163.574526           1  
2   225.412882  139766.247529     32163.574526           0  
3   225.412882  139766.247529     32163.574526           0  
4   225.412882  139766.247529     32163.574526           0  

[5 rows x 53 columns]

**Removing the logically unimportant variables**

1) policy_code as all the values are '1'
2) pymnt_plan as all the values are 'n'
3) 'id' and 'member_id' : there should be a unique id but in this case there are 2 id's creating confusion
4) 'initial_list_status' as it not an important factor
5) acc_now_delinq as it has similar value '0' for more than 99%
6) emp_title and title
7) zip_code doest not matter for loan qualification
8) sub_grade as there exixts a similar variable called 'grade' already
9) pub_rec as public record is not logically important
10) addr_state as address is not important for loan
11) inq_last_6mths and open_acc are not important factors

In [24]:
removable_vars = ['policy_code','id', 'member_id', 'addr_state', 'initial_list_status', 'pymnt_plan', 'open_acc',
                      'acc_now_delinq','emp_title','sub_grade', 'pub_rec', 'zip_code','title','inq_last_6mths']

df.drop( removable_vars , axis = 1, inplace = True )

In [25]:
df.columns

Index(['loan_amnt', 'funded_amnt', 'funded_amnt_inv', 'term', 'int_rate',
       'installment', 'grade', 'emp_length', 'home_ownership', 'annual_inc',
       'verification_status', 'issue_d', 'purpose', 'dti', 'delinq_2yrs',
       'earliest_cr_line', 'mths_since_last_delinq', 'revol_bal', 'revol_util',
       'total_acc', 'out_prncp', 'out_prncp_inv', 'total_pymnt',
       'total_pymnt_inv', 'total_rec_prncp', 'total_rec_int',
       'total_rec_late_fee', 'recoveries', 'collection_recovery_fee',
       'last_pymnt_d', 'last_pymnt_amnt', 'next_pymnt_d', 'last_credit_pull_d',
       'collections_12_mths_ex_med', 'application_type', 'tot_coll_amt',
       'tot_cur_bal', 'total_rev_hi_lim', 'default_ind'],
      dtype='object')

In [26]:
df.shape

(855969, 39)

In [ ]:
df.dtypes

# Data Visualization

In [ ]:
import seaborn as sns

In [ ]:
sns.countplot(data=df, x='home_ownership',palette='Set1')
plt.show()

In [ ]:
import seaborn as sns
sns.countplot(x='default_ind',data=df,palette='Set2') 

In [ ]:
sns.countplot(data=data,x='application_type')

plt.show()

In [ ]:
sns.countplot(x='grade',data=df) 

In [ ]:
sns.countplot(x='purpose',data=df) 
plt.xticks(rotation=90)

**Finding the object variables in the dataframe using 'for' loop**

In [ ]:
colname = []

for x in df.columns[:]:
    if df[x].dtype=='object':
        colname.append(x)
        
colname


**After getting a list of all object variables we notice that the variable 'issuue_d' is also present in the list**

**As we need 'issue_d' in the next few steps to split data we will exclude it from the list to further convert it into a datetime variable**

In [27]:
colname = ['term',
 'grade',
 'emp_length',
 'home_ownership',
 'verification_status',
 'purpose',
 'earliest_cr_line',
 'last_pymnt_d',
 'next_pymnt_d',
 'last_credit_pull_d',
 'application_type']

In [28]:
# For Preprocessing the data
# convert all categorical variables into numeric variables

from sklearn import preprocessing

le=preprocessing.LabelEncoder()    #LabelEncoder()  == technique of transformation == object to int/float

for x in colname:
    df[x]=le.fit_transform(df[x])

In [ ]:
df.dtypes

Convering 'issue_d' into datetime for splitting the data

In [29]:
df.issue_d = pd.to_datetime(df.issue_d)
col_name = 'issue_d'
print(df[col_name].dtype)

datetime64[ns]


**Split the data into train and test according to the condition given in the problem statement**

In [30]:
split_date = "2015-06-01"

XYZ_training = df[df['issue_d']<split_date]
XYZ_training = XYZ_training.drop(['issue_d'],axis=1)

XYZ_test = df[df['issue_d'] >= split_date]
XYZ_test = XYZ_test.drop(['issue_d'],axis=1)
print(XYZ_test.shape)

print(XYZ_training.shape)

(256991, 38)
(598978, 38)


In [31]:
X_train=XYZ_training.values[:,:-1]
Y_train=XYZ_training.values[:,-1]
Y_train=Y_train.astype(int)
print(Y_train)

X_test=XYZ_test.values[:,:-1]
Y_test=XYZ_test.values[:,-1]
Y_test=Y_test.astype(int)
print(Y_test)

[0 1 0 ... 0 0 0]
[0 0 0 ... 0 0 0]


### Scaling X_train and X_test for model building
**Scaling is not compulsory but can be used to get better accuracy**

**Scaling can be done only on X**

**Standardization is used mostly for better accuracy**

**Fitting is only done on training data**

In [32]:
from sklearn.preprocessing import StandardScaler

scaler = StandardScaler()

scaler.fit(X_train)

X_train = scaler.transform(X_train)
X_test = scaler.transform(X_test)
print(XYZ_training)


        loan_amnt  funded_amnt  funded_amnt_inv  term  int_rate  installment  \
0          5000.0       5000.0      4975.000000     0     10.65       162.87   
1          2500.0       2500.0      2500.000000     1     15.27        59.83   
2          2400.0       2400.0      2400.000000     0     15.96        84.33   
3         10000.0      10000.0     10000.000000     0     13.49       339.31   
4          3000.0       3000.0      3000.000000     1     12.69        67.79   
5          5000.0       5000.0      5000.000000     0      7.90       156.46   
6          7000.0       7000.0      7000.000000     1     15.96       170.08   
7          3000.0       3000.0      3000.000000     0     18.64       109.43   
8          5600.0       5600.0      5600.000000     1     21.28       152.39   
9          5375.0       5375.0      5350.000000     1     12.69       121.45   
10         6500.0       6500.0      6500.000000     1     14.65       153.45   
11        12000.0      12000.0     12000

# Model Building

### Logistic Regression

In [ ]:
from sklearn.linear_model import LogisticRegression

#create a model object
classifier=LogisticRegression()

#fitting training data to the model
classifier.fit(X_train,Y_train)

Y_pred=classifier.predict(X_test)
print(list(zip(Y_test,Y_pred)))

print(classifier.coef_)
print(classifier.intercept_)

In [ ]:
from sklearn.metrics import confusion_matrix,accuracy_score,classification_report
cfm=confusion_matrix(Y_test,Y_pred)
print(cfm)

print("classification report: ")
print(classification_report(Y_test,Y_pred))
acc=accuracy_score(Y_test,Y_pred)
print("accuracy of the model:",acc)

### Cross Validation Model

In [ ]:
classifier=(LogisticRegression())    #pass the name of the parameter you want to run

#performing kfold_cross_validation
from sklearn.model_selection import KFold
kfold_cv=KFold(n_splits=10)      #10 iterations can be changed also, by default 3
print(kfold_cv)

from sklearn.model_selection import cross_val_score
#running the model using scoring metric as accuracy
kfold_cv_result=cross_val_score(estimator=classifier,X=X_train,y=Y_train, cv=kfold_cv)
#estimator == specify the name of the algorithm (logistic), 10 fold model on training data

print(kfold_cv_result)
#finding the mean
print(kfold_cv_result.mean())


In [ ]:
for train_value, test_value in kfold_cv.split(X_train):
    classifier.fit(X_train[train_value], Y_train[train_value]).predict(X_train[test_value])
    
#train_value represents 9 train folds and test_value represents the 1 test fold

Y_pred=classifier.predict(X_test)
#print(list(zip(Y_test,Y_pred)))


In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report


cfm=confusion_matrix(Y_test,Y_pred)
print(cfm)

print("Classification report: ")


print(classification_report(Y_test,Y_pred))

acc=accuracy_score(Y_test,Y_pred)
print("Accuracy of the model: ",acc)


### Adjusting the threshold

In [ ]:
#store the predicted probabilities

y_pred_prob = classifier.predict_proba(X_test)
print(y_pred_prob)

In [ ]:
y_pred_class = []    #create an empty list

for value in y_pred_prob[:,1]:       #all rows but only 1st column
    if value > 0.45:
        y_pred_class.append(1)
    else:
        y_pred_class.append(0)
print(y_pred_class)

In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report


cfm=confusion_matrix(Y_test,y_pred_class)
print(cfm)

print("Classification report: ")


print(classification_report(Y_test,y_pred_class))

acc=accuracy_score(Y_test,y_pred_class)
print("Accuracy of the model: ",acc)



**Running a Generic Code to check the best value of threshold for the lowest amaount of errors**

In [ ]:
for a in np.arange(0,1,0.05):
    predict_mine = np.where(y_pred_prob[:,1] > a, 1, 0)
    cfm=confusion_matrix(Y_test, predict_mine)
    total_err=cfm[0,1]+cfm[1,0]     #first row and 0th index, 2nd row and 1st index
    print("Errors at threshold ", a, ":",total_err, " , type 2 error :", 
        cfm[1,0]," , type 1 error:", cfm[0,1])

In [ ]:
y_pred_class = []    #create an empty list

for value in y_pred_prob[:,1]:       #all rows but only 1st column
    if value > 0.95:
        y_pred_class.append(1)
    else:
        y_pred_class.append(0)
print(y_pred_class)

In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report


cfm=confusion_matrix(Y_test,y_pred_class)
print(cfm)

print("Classification report: ")


print(classification_report(Y_test,y_pred_class))

acc=accuracy_score(Y_test,y_pred_class)
print("Accuracy of the model: ",acc)



### ROC Curve

In [ ]:
from sklearn import metrics

fpr, tpr,z = metrics.roc_curve(Y_test, y_pred_class)     #for threshold 0.70
auc = metrics.auc(fpr,tpr)
print(auc)
print(fpr)
print(tpr)

In [ ]:
import matplotlib.pyplot as plt
#%matplotlib inline
plt.title('Receiver Operating Characteristic')    #title
plt.plot(fpr, tpr, 'b', label = auc)    #b=color of auc line, label = auc
plt.legend(loc = 'lower right')     #legend=to place auc value according to user
plt.plot([0, 1], [0, 1],'r--')     #'r--'=red dotted line, ([0,1],[0,1]) = x and y axis
plt.xlim([0, 1])
plt.ylim([0, 1])
plt.xlabel('False Positive Rate')
plt.ylabel('True Positive Rate')

plt.show()



### Decision Tree

In [33]:
from sklearn.tree import DecisionTreeClassifier

model_DecisionTree=DecisionTreeClassifier(random_state=10,min_samples_leaf=10,max_depth=10)

#fit the modl on the data predit the values
model_DecisionTree.fit(X_train,Y_train)

DecisionTreeClassifier(class_weight=None, criterion='gini', max_depth=10,
            max_features=None, max_leaf_nodes=None,
            min_impurity_decrease=0.0, min_impurity_split=None,
            min_samples_leaf=10, min_samples_split=2,
            min_weight_fraction_leaf=0.0, presort=False, random_state=10,
            splitter='best')

In [34]:
Y_pred=model_DecisionTree.predict(X_test)
print(Y_pred)
#print(list(zip(Y_test,Y_pred)))

[0 0 0 ... 0 0 0]


In [35]:
##confusion matrix and accuracy
from sklearn.metrics import confusion_matrix,accuracy_score,\
classification_report
cfm=confusion_matrix(Y_test,Y_pred)
print(cfm)

print("classification report: ")
print(classification_report(Y_test,Y_pred))

acc=accuracy_score(Y_test,Y_pred)
print("accuracy report:",acc)


[[256676      4]
 [    10    301]]
classification report: 
              precision    recall  f1-score   support

           0       1.00      1.00      1.00    256680
           1       0.99      0.97      0.98       311

   micro avg       1.00      1.00      1.00    256991
   macro avg       0.99      0.98      0.99    256991
weighted avg       1.00      1.00      1.00    256991

accuracy report: 0.9999455233840874


In [36]:
XYZ_test["Y_Predicted"] = Y_pred

In [ ]:
export_df= XYZ_test.to_excel(r'C:\Python\PythonProject-1\Python Project - Bank Lending\Y_PRED.xlsx',index = None, header=True)

### ROC Curve

In [ ]:
from sklearn import metrics

fpr, tpr,z = metrics.roc_curve(Y_test, Y_pred)     #for default threshold 0.5
auc = metrics.auc(fpr,tpr)
print(auc)
print(fpr)
print(tpr)

In [ ]:
import matplotlib.pyplot as plt
#%matplotlib inline
plt.title('Receiver Operating Characteristic')    #title
plt.plot(fpr, tpr, 'b', label = auc)    #b=color of auc line, label = auc
plt.legend(loc = 'lower right')     #legend=to place auc value according to user
plt.plot([0, 1], [0, 1],'r--')     #'r--'=red dotted line, ([0,1],[0,1]) = x and y axis
plt.xlim([0, 1])
plt.ylim([0, 1])
plt.xlabel('False Positive Rate')
plt.ylabel('True Positive Rate')

plt.show()



In [ ]:
#store the predicted probabilities

y_pred_prob1 = model_DecisionTree.predict_proba(X_test)
print(y_pred_prob1)

In [ ]:
for a in np.arange(0,1,0.05):
    predict_mine = np.where(y_pred_prob1[:,1] > a, 1, 0)
    cfm=confusion_matrix(Y_test, predict_mine)
    total_err=cfm[0,1]+cfm[1,0]     #first row and 0th index, 2nd row and 1st index
    print("Errors at threshold ", a, ":",total_err, " , type 2 error :", 
        cfm[1,0]," , type 1 error:", cfm[0,1])

In [ ]:
classifier=(DecisionTreeClassifier())    #pass the name of the parameter you want to run

#performing kfold_cross_validation
from sklearn.model_selection import KFold
kfold_cv=KFold(n_splits=5)      #5 iterations can be changed also, by default 3
print(kfold_cv)

from sklearn.model_selection import cross_val_score
#running the model using scoring metric as accuracy
kfold_cv_result=cross_val_score(estimator=classifier,X=X_train,y=Y_train, cv=kfold_cv)
#estimator == specify the name of the algorithm (logistic), 10 fold model on training data

print(kfold_cv_result)
#finding the mean
print(kfold_cv_result.mean())


In [ ]:
for train_value, test_value in kfold_cv.split(X_train):
    classifier.fit(X_train[train_value], Y_train[train_value]).predict(X_train[test_value])
    
#train_value represents 4 train folds and test_value represents the 1 test fold

Y_pred=classifier.predict(X_test)
#print(list(zip(Y_test,Y_pred)))


In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report


cfm=confusion_matrix(Y_test,Y_pred)
print(cfm)

print("Classification report: ")


print(classification_report(Y_test,Y_pred))

acc=accuracy_score(Y_test,Y_pred)
print("Accuracy of the model: ",acc)


In [ ]:
classifier=(DecisionTreeClassifier())

In [ ]:
colname=df.columns[:]

In [ ]:
from sklearn.feature_selection import RFE

rfe = RFE(classifier,10)
model_rfe = rfe.fit(X_train,Y_train)

print("Num Features: ",model_rfe.n_features_)     #number of iterations
print("Selected Features: ")
print(list(zip(colname,model_rfe.support_)))     #boolean list showing which elements were retained and which were eliminated
print("Feature Ranking: ",model_rfe.ranking_)    #rank 1 == retained elements, highest rank means the element was eliminated first


In [ ]:
y_pred_rfe=model_rfe.predict(X_test)

In [ ]:
from sklearn.metrics import confusion_matrix, accuracy_score, classification_report


cmatrix=confusion_matrix(Y_test,y_pred_rfe)
print(cmatrix)

print("Classification report: ")


print(classification_report(Y_test,y_pred_rfe))

acc2=accuracy_score(Y_test,y_pred_rfe)
print("Accuracy of the model: ",acc2)


### ExtraTreesClassifier

In [ ]:
#Running ExtraTreesClassifier model
#prediction using the Bagging_classifier

from sklearn.ensemble import ExtraTreesClassifier

model=ExtraTreesClassifier(50,random_state=10)
#fit the model on the data and predict the values
model=model.fit(X_train,Y_train)

Y_pred=model.predict(X_test)

In [ ]:
from sklearn.metrics import confusion_matrix,accuracy_score,classification_report

cfm=confusion_matrix(Y_test,Y_pred)
print(cfm)

print('Classification report')

print(classification_report(Y_test,Y_pred))

acc=accuracy_score(Y_test,Y_pred)
print('Accuracy of the model:',acc)

### AdaBoost_Classifier

In [ ]:
#prediction using the AdaBoost_Classifier
from sklearn.ensemble import AdaBoostClassifier

model_AdaBoost=AdaBoostClassifier(base_estimator=DecisionTreeClassifier(),n_estimators=50,random_state=10)
#fit the model on the data and predict the values
model_AdaBoost.fit(X_train,Y_train)

Y_pred=model_AdaBoost.predict(X_test)

In [ ]:
from sklearn.metrics import confusion_matrix,accuracy_score,classification_report

cfm=confusion_matrix(Y_test,Y_pred)
print(cfm)

print('Classification report')

print(classification_report(Y_test,Y_pred))

acc=accuracy_score(Y_test,Y_pred)
print('Accuracy of the model:',acc)

### Gradient_Boosting_Classifier

In [ ]:
#predicting using the Gradient_Boosting_Classifier

from sklearn.ensemble import GradientBoostingClassifier

model_gb= GradientBoostingClassifier(n_estimators=100,random_state=10)

#fit the model on the data and predict the calues

model_gb = model_gb.fit(X_train,Y_train)

Y_test = model_gb.predict(X_test)

In [ ]:
#evaluate the model

from sklearn.metrics import confusion_matrix, accuracy_score, classification_report


cfm=confusion_matrix(Y_test,Y_pred)
print(cfm)

print("Classification report: ")


print(classification_report(Y_test,Y_pred))

acc=accuracy_score(Y_test,Y_pred)
print("Accuracy of the model: ",acc)

### RandomForestClassifier

In [ ]:
from sklearn.ensemble import RandomForestClassifier

model_RandomForest=RandomForestClassifier(100,random_state=10)
#fit the model on the data and predict the values
model_RandomForest.fit(X_train,Y_train)

Y_pred=model_RandomForest.predict(X_test)

In [ ]:
from sklearn.metrics import confusion_matrix,accuracy_score,classification_report

cfm=confusion_matrix(Y_test,Y_pred)
print(cfm)

print('Classification report')

print(classification_report(Y_test,Y_pred))

acc=accuracy_score(Y_test,Y_pred)
print('Accuracy of the model:',acc)

In [37]:
# tabulation

print(XYZ_test.default_ind.value_counts())

0    256680
1       311
Name: default_ind, dtype: int64


In [38]:
# tabulation

print(XYZ_test.Y_Predicted.value_counts())

0    256686
1       305
Name: Y_Predicted, dtype: int64


In [ ]:
XYZ_test['Y_PRED']=Y_pred

In [ ]:
XYZ_test.columns

In [ ]:
# tabulation

print(df.default_ind.value_counts())

In [ ]:
models_report_oot

In [ ]:
from sklearn.linear_model import LogisticRegression
from sklearn.tree import DecisionTreeClassifier
from sklearn.ensemble import VotingClassifier

 

# create the sub models
estimators = []
model1 = LogisticRegression()
estimators.append(('log', model1))
model2 = DecisionTreeClassifier(criterion='gini',random_state=10)
estimators.append(('cart', model2))

# create the ensemble model
ensemble = VotingClassifier(estimators)
ensemble.fit(X_train,Y_train)
Y_pred=ensemble.predict(X_test)
#print(Y_pred)

In [ ]:
from sklearn.metrics import confusion_matrix,accuracy_score,classification_report

cfm=confusion_matrix(Y_test,Y_pred)
print(cfm)

print('Classification report')

print(classification_report(Y_test,Y_pred))

acc=accuracy_score(Y_test,Y_pred)
print('Accuracy of the model:',acc)